Decode and treat as a certain failure all non-nonce error responses

itsmojo opened the pull request:
Along with rileylink_ios PR #571, remaining portion of fix for issue #569:
Pod non badNonce Error Responses are not decoded, logged or treated as a certain failure

itsmojo committed:
Decode and treat as a certain failure all non-nonce error responses
Along with rileylink_ios PR #571, remaining portion of fix for issue #569:
Pod non badNonce Error Responses are not decoded, logged or treated as a certain failure

ps2 commented on OmniKit/MessageTransport/MessageBlocks/ErrorResponse.swift:
```diff
@@ -8,16 +8,16 @@
     
     public let blockType: MessageBlockType = .errorResponse
     
-    public enum ErrorReponseType: UInt8 {
```
Looks like this should rather be an enum with associated types:
```
public enum ErrorResponseType {
    case badNonce(nonceSearchKey: UInt16)
    case fault(faultEventCode: FaultEventCode, progress: PodProgressStatus)
}

itsmojo replied:
Completely reworked to use a enum case with parameters

itsmojo committed:
Reworked ErrorResponse to use enum case with parameters
Add a new test case for a non bad nonce error response
Treat a PodCommsError.commandError as a certain failure
Updated commenting and send() logic

itsmojo committed:
Comment out failureReason definition to fix merge conflict

itsmojo committed:
Update PodCommsError LocalizedError commandError description string to match

itsmojo commented:
On May 12, 2020, at 2:47 PM, Pete Schwamb <notifications@github.com> wrote:

*@ps2* requested changes on this pull request.
------------------------------

In OmniKit/MessageTransport/MessageBlocks/ErrorResponse.swift
<https://github.com/ps2/rileylink_ios/pull/596#discussion_r424052937>:

>  public struct ErrorResponse : MessageBlock {

     public let blockType: MessageBlockType = .errorResponse

-    public enum ErrorReponseType: UInt8 {

Looks like this should rather be an enum with associated types:

public enum ErrorResponseType {
    case badNonce(nonceSearchKey: UInt16)
    case fault(faultEventCode: FaultEventCode, progress: PodProgressStatus)
}

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/ps2/rileylink_ios/pull/596#pullrequestreview-410446503>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AFPUETYQ627Q7CX6HRX5TQLRRG7XDANCNFSM4MGRMREA>
.

I tried doing something like this originally and ran into some difficulties
because I am not a skilled Swift programmer. I spent the time to do it this
way yesterday.  It caused me to end up reworking some of the error logic in
send(), but I think the result is an improvement and it was all worth it.
It also got me to add an additional OmnikitTest test for a non bad nonce
ErrorResponse (which I should have had before) and to got me to realize
that I needed to add code to make sure error these cases are actually
treated as certain failures for things like bolus & cancel.

For my testing of this stuff yesterday, I added a GetStatusCall() to
PodCommSession’s bolus() call to send() (making it 3 have 3 sub commands)
and manually bolused to create error returns (the Insulin commands insist
that the there has to exactly 2 sub commands or else they return
ErrorResponse 7). I typically put my pod in and out of the faraday bag to
force nonce resync conditions via ErrorResponse 0x14.

itsmojo committed:
Remove unneeded & incorrect errorResponse handling in sendMessage()

itsmojo committed:
Merge branch 'dev' into decode-all-error-responses

itsmojo committed:
Merge branch 'dev' into decode-all-error-responses

itsmojo committed:
Update PodCommsSession.swift

Formatting clean up

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -28,6 +28,7 @@ public enum PodCommsError: Error {
     case podSuspended
     case podFault(fault: PodInfoFaultEvent)
     case commsError(error: Error)
+    case commandError(errorCode: UInt8)
```
Maybe `rejectedCommand`, or `unretryableCommand`? There are many errors that can happen when sending a command that don't result in a "commandError" so I think this is too broad. Would be good if the name indicated something about the kind of error the command caused.

ps2 reviewed (approved):
LGTM. I think it would be good if the new `commandError` error type were more specific.

itsmojo committed:
Update PodCommsSession.swift

Change PodCommsError.commandError to more specific invalidMessage

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -190,7 +190,7 @@ public class PodCommsSession {
     ///     - PodCommsError.podFault
     ///     - PodCommsError.unexpectedResponse
-    ///     - PodCommsError.commandError
+    ///     - PodCommsError.illegalMessage
```
```suggestion
    ///     - PodCommsError.invalidMessage
```
I did like rejectedMessage a bit better; invalid sounds like something that fails our idea of invalid and might be an error our comms layer could generate, whereas 'rejected' communicates it failed some idea of validity on the recipient's side.

ps2 reviewed (approved):
LGTM. Just a comment typo, and another thought on naming, that's not critical; I'd accept "invalidMessage"

itsmojo committed:
Use rejectedMessage for new PodCommsError error
